--- Fast.pm.~7~	2005-01-11 10:02:04.000000000 -0800
+++ Fast.pm	2005-04-19 20:38:58.000000000 -0700
@@ -558,6 +558,7 @@
 		      Line      => $ln,
 		      name      => $domain,
 		      type      => "RRSIG",
+		      class     => "IN",
 		      ttl       => $ttl,
 		      typecovered => $1,
 		      algorithm => $2,
@@ -587,6 +588,8 @@
 			   first     => 1,
 			   Line      => $ln,
 			   name      => $domain,
+			   ttl       => $ttl,
+			   class     => "IN",
 			   type      => "DNSKEY",
 			   flags     => $1,
 			   protocol  => $2,
@@ -612,6 +615,8 @@
 		$ds = {
 		       Line      => $ln,
 		       name      => $domain,
+		       class     => "IN",
+		       ttl       => $ttl,
 		       type      => "DS",
 		       keytag    => $1,
 		       algorithm => $2,
@@ -637,6 +642,8 @@
 		      {
 		       Line      => $ln,
 		       name      => $domain,
+		       class     => "IN",
+		       ttl       => $ttl,
 		       type      => "NSEC",
 		       nxtdname  => lc($1),
 		       typelist  => $2
