Index: validator/libval_shim/Makefile.in
===================================================================
--- validator/libval_shim/Makefile.in	(revision 5149)
+++ validator/libval_shim/Makefile.in	(working copy)
@@ -1,6 +1,6 @@
 
 INCL= ../include
-EXTRA_LIBS= -L../libsres/.libs -lsres -L../libval/.libs -lval$(LIBVAL_SUFFIX) -lpthread @LDFLAGS@ -lcrypto
+EXTRA_LIBS=../libsres/libsres.la ../libval/libval$(LIBVAL_SUFFIX).la -lpthread @LDFLAGS@ -lcrypto
 CFLAGS_EX=-I$(INCL) -I. -I../libsres -I../libval -D_GNU_SOURCE 
 
 SRC=  	libval_shim.c 
