From 92bdd67b4834bb1d89874d373f3a5a350641b300 Mon Sep 17 00:00:00 2001
From: Robert Story <rstory@localhost>
Date: Sat, 22 Dec 2012 17:56:25 -0500
Subject: [PATCH 18/22] DANE move new function inside braces

---
 security/nss/lib/ssl/ssl.def | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/nss/lib/ssl/ssl.def b/security/nss/lib/ssl/ssl.def
index 6f24075..b60cf90 100644
--- a/security/nss/lib/ssl/ssl.def
+++ b/security/nss/lib/ssl/ssl.def
@@ -138,6 +138,7 @@ SSL_SetNextProtoCallback;
 SSL_SetNextProtoNego;
 SSL_GetNextProto;
 SSL_AuthCertificateComplete;
+SSL_DANECheck;
 ;+    local:
 ;+       *;
 ;+};
-- 
1.7.11.7

