diff -r 6d8bcd20f245 mail/components/preferences/advanced.xul
--- a/mail/components/preferences/advanced.xul	Tue May 31 01:56:15 2011 +0800
+++ b/mail/components/preferences/advanced.xul	Tue May 31 11:07:41 2011 -0400
@@ -141,6 +141,9 @@
       <preference id="security.disable_button.openDeviceManager"
                   name="security.disable_button.openDeviceManager"
                   type="bool"/>
+      <preference id="security.dnssec.dnssecBehavior"
+                  name="security.dnssec.dnssecBehavior"
+                  type="int"/>
     </preferences>
 
 #ifdef HAVE_SHELL_SERVICE
@@ -155,6 +158,7 @@
         <tab id="networkingTab" label="&itemNetworking.label;"/>
         <tab label="&itemUpdate.label;"/>
         <tab label="&itemCertificates.label;"/>
+        <tab label="&itemDNSSEC.label;"/>
       </tabs>
 
       <tabpanels flex="1">
@@ -428,6 +432,19 @@
           </vbox>
 #endif
         </tabpanel>
+        <!-- DNS Security --> 
+        <tabpanel orient="vertical">
+          <description>&dnssec.description;</description>
+          <separator class="thin"/>
+          <groupbox>
+            <radiogroup id="dnssecBehaviour" preference="security.dnssec.dnssecBehavior">
+                <radio value="1" label="&neverDNSSecure.label;"
+                       accesskey="&neverDNSSecure.accesskey;"/>
+                <radio value="0" label="&possibleDNSSecure.label;"
+                       accesskey="&possibleDNSSecure.accesskey;"/>
+            </radiogroup>
+          </groupbox>
+        </tabpanel>
       </tabpanels>
     </tabbox>
     <separator/>
diff -r 6d8bcd20f245 mail/locales/en-US/chrome/messenger/preferences/advanced.dtd
--- a/mail/locales/en-US/chrome/messenger/preferences/advanced.dtd	Tue May 31 01:56:15 2011 +0800
+++ b/mail/locales/en-US/chrome/messenger/preferences/advanced.dtd	Tue May 31 11:07:41 2011 -0400
@@ -106,3 +106,11 @@
 <!ENTITY validation.accesskey           "V">
 <!ENTITY viewSecurityDevices.label      "Security Devices">
 <!ENTITY viewSecurityDevices.accesskey  "S">
+
+<!-- DNS Security -->
+<!ENTITY itemDNSSEC.label               "DNS Security">
+<!ENTITY dnssec.description             "Enforce or Ignore System DNS Security Policy.">
+<!ENTITY neverDNSSecure.label           "Ignore DNS security">
+<!ENTITY neverDNSSecure.accesskey       "N">
+<!ENTITY possibleDNSSecure.label        "Enforce DNS Security">
+<!ENTITY possibleDNSSecure.accesskey    "E">
