--- .mozconfig.dnssec	2009-12-04 00:36:33.708905113 -0500
+++ .mozconfig	2009-12-04 00:41:01.457845551 -0500
@@ -10,7 +10,7 @@
 ac_add_options --with-pthreads
 ac_add_options --disable-tests
 ac_add_options --disable-debug
-ac_add_options --enable-optimize="$RPM_OPT_FLAGS"
+ac_add_options --enable-optimize="$MOZ_OPT_FLAGS"
 #ac_add_options --enable-debug
 #ac_add_options --disable-optimize
 ac_add_options --disable-installer
