--- mozilla.firefox.1.5.0.6.rpmdist/browser/components/preferences/advanced.xul	2005-09-14 14:24:40.000000000 -0700
+++ mozilla.newdnssec/browser/components/preferences/advanced.xul	2006-08-23 12:10:47.000000000 -0700
@@ -94,6 +94,9 @@
       <preference id="security.disable_button.openDeviceManager"
                   name="security.disable_button.openDeviceManager"
                   type="bool"/>
+      <preference id="security.dnssec.dnssecBehavior"
+                  name="security.dnssec.dnssecBehavior"
+                  type="int"/>
     </preferences>
     
     <script type="application/x-javascript" src="chrome://browser/content/preferences/advanced.js"/>
@@ -104,6 +107,7 @@
         <tab label="&generalTab.label;" helpTopic="prefs-advanced-general"/>
         <tab label="&updateTab.label;" helpTopic="prefs-advanced-update"/>
         <tab label="&securityTab.label;" helpTopic="prefs-advanced-security"/>
+        <tab label="&dnssecTab.label;" helpTopic="prefs-advanced-dnssec"/>
       </tabs>
       <tabpanels flex="1">
         <tabpanel orient="vertical">
@@ -251,6 +255,21 @@
 #endif
           </groupbox>
         </tabpanel>
+        <tabpanel orient="vertical">
+          <groupbox>
+            <caption label="&dnsSecurity.label;"/>
+            <description>&dnssec.description;</description>
+	    <radiogroup id="networkDnssecBehaviour"
+	                preference="security.dnssec.dnssecBehavior">
+	      <radio value="0" label="&neverDNSSecure.label;"
+	             accesskey="&neverDNSSecure.accesskey;"/>
+	      <radio value="1" label="&possibleDNSSecure.label;"
+	             accesskey="&possibleDNSSecure.accesskey;"/>
+	      <radio value="2" label="&allDNSSecure.label;"
+	             accesskey="&allDNSSecure.accesskey;"/>
+	    </radiogroup>
+          </groupbox>
+        </tabpanel>
       </tabpanels>
     </tabbox>   
   </prefpane>
--- mozilla.firefox.1.5.0.6.rpmdist/browser/locales/en-US/chrome/browser/preferences/advanced.dtd	2005-10-14 19:57:43.000000000 -0700
+++ mozilla.newdnssec/browser/locales/en-US/chrome/browser/preferences/advanced.dtd	2006-08-23 11:41:37.000000000 -0700
@@ -56,6 +56,16 @@
 <!ENTITY viewSecurityDevices.label      "Security Devices">
 <!ENTITY viewSecurityDevices.accesskey  "y">
 
+<!ENTITY dnssecTab.label                "DNS Security">
+<!ENTITY dnssec.description             "When should Server Name Lookups be Secure:">
+<!ENTITY dnsSecurity.label              "DNS Security Options">
+<!ENTITY neverDNSSecure.label           "Ignore DNS security">
+<!ENTITY neverDNSSecure.accesskey       "N">
+<!ENTITY possibleDNSSecure.label        "Use DNS Security if possible">
+<!ENTITY possibleDNSSecure.accesskey    "P">
+<!ENTITY allDNSSecure.label            "Require all DNS lookups to be secure">
+<!ENTITY allDNSSecure.accesskey         "A">
+
 <!ENTITY languages.caption              "Languages">
 <!ENTITY languagesInfo.label            "Choose Languages web pages are displayed in.">
 <!ENTITY showLanguages.label            "Edit Languages...">
--- mozilla.firefox.1.5.0.6.rpmdist/browser/locales/en-US/chrome/overrides/appstrings.properties	2005-09-02 14:39:52.000000000 -0700
+++ mozilla.newdnssec/browser/locales/en-US/chrome/overrides/appstrings.properties	2006-08-22 16:47:31.000000000 -0700
@@ -37,6 +37,7 @@
 malformedURI=The URL is not valid and cannot be loaded.
 fileNotFound=Firefox can't find the file at %S.
 dnsNotFound=Firefox can't find the server at %S.
+dnsDNSSECError=%S failed its DNSSEC security check validation. Please contact a security or system administrator for help.
 protocolNotFound=Firefox doesn't know how to open this address, because the protocol (%S) isn't associated with any program.
 connectionFailure=Firefox can't establish a connection to the server at %S.
 netInterrupt=The connection to %S was interrupted while the page was loading.
--- mozilla.firefox.1.5.0.6.rpmdist/browser/locales/en-US/chrome/overrides/netError.dtd	2005-08-29 13:10:06.000000000 -0700
+++ mozilla.newdnssec/browser/locales/en-US/chrome/overrides/netError.dtd	2006-08-22 16:47:31.000000000 -0700
@@ -25,6 +25,9 @@
 </ul>
 ">
 
+<!ENTITY dnsDNSSECError.title "DNSSEC Validation Error">
+<!ENTITY dnsDNSSECError.longDesc "The address (URL) failed its DNSSEC security check validation.  Please contact a system administrator for help.">
+
 <!ENTITY fileNotFound.title "File not found">
 <!ENTITY fileNotFound.longDesc "
 <ul>
--- mozilla.firefox.1.5.0.6.rpmdist/dom/locales/en-US/chrome/netError.dtd	2005-08-21 20:23:35.000000000 -0700
+++ mozilla.newdnssec/dom/locales/en-US/chrome/netError.dtd	2006-08-22 16:47:31.000000000 -0700
@@ -12,6 +12,9 @@
 <!ENTITY dnsNotFound.title "Address Not Found">
 <!ENTITY dnsNotFound.longDesc "<p>The browser could not find the host server for the provided address.</p><ul><li>Did you make a mistake when typing the domain? (e.g. <q><strong>ww</strong>.mozilla.org</q> instead of <q><strong>www</strong>.mozilla.org</q>)</li><li>Are you certain this domain address exists?  Its registration may have expired.</li><li>Are you unable to browse other sites?  Check your network connection and DNS server settings.</li><li>Is your computer or network protected by a firewall or proxy?  Incorrect settings can interfere with Web browsing.</li></ul>">
 
+<!ENTITY dnsDNSSECError.title "DNSSEC Validation Error">
+<!ENTITY dnsDNSSECError.longDesc "The address (URL) failed its DNSSEC security check validation.  Please contact a system administrator for help.">
+
 <!ENTITY fileNotFound.title "File Not Found">
 <!ENTITY fileNotFound.longDesc "<ul><li>Could the item have been renamed, removed, or relocated?</li><li>Is there a spelling, capitalization, or other typographical error in the address?</li><li>Do you have sufficient access permissions to the requested item?</li></ul>">
 
--- mozilla.firefox.1.5.0.6.rpmdist/layout/base/nsPresContext.cpp	2006-04-21 16:30:50.000000000 -0700
+++ mozilla.newdnssec/layout/base/nsPresContext.cpp	2006-08-23 13:18:17.000000000 -0700
@@ -68,6 +68,7 @@
 #include "nsIWeakReferenceUtils.h"
 #include "nsCSSRendering.h"
 #include "prprf.h"
+#include "prnetdb.h"
 #include "nsContentPolicyUtils.h"
 #include "nsIScriptGlobalObject.h"
 #include "nsIDOMDocument.h"
@@ -258,6 +259,14 @@
   nsContentUtils::UnregisterPrefCallback("image.animation_mode",
                                          nsPresContext::PrefChangedCallback,
                                          this);
+  nsContentUtils::UnregisterPrefCallback("image.animation_mode",
+                                         nsPresContext::PrefChangedCallback,
+                                         this);
+#ifdef MOZ_DNSSEC
+  nsContentUtils::UnregisterPrefCallback("security.dnssec.dnssecBehavior",
+                                         nsPresContext::PrefChangedCallback,
+                                         this);
+#endif
 #ifdef IBMBIDI
   nsContentUtils::UnregisterPrefCallback("bidi.", PrefChangedCallback, this);
 
@@ -571,6 +580,17 @@
   else if (animatePref.Equals("once"))
     mImageAnimationModePref = imgIContainer::kLoopOnceAnimMode;
 
+  fprintf(stderr, "XYZhere : before\n");
+#ifdef MOZ_DNSSEC
+  {
+    int dnssecpref =
+      nsContentUtils::GetIntPref("security.dnssec.dnssecBehavior");
+    fprintf(stderr, "XYZhere : yes: %d\n", dnssecpref);
+    PR_set_dnssec_validate_policy(dnssecpref);
+  }
+#endif /* MOZ_DNSSEC */
+
+
 #ifdef IBMBIDI
   PRInt32 prefInt =
     nsContentUtils::GetIntPref("bidi.direction",
@@ -704,6 +724,13 @@
   nsContentUtils::RegisterPrefCallback("image.animation_mode",
                                        nsPresContext::PrefChangedCallback,
                                        this);
+  fprintf(stderr, "ABChere out:\n");
+#ifdef MOZ_DNSSEC
+  fprintf(stderr, "ABChere in:\n");
+  nsContentUtils::RegisterPrefCallback("security.dnssec.dnssecBehavior",
+                                       nsPresContext::PrefChangedCallback,
+                                       this);
+#endif
 #ifdef IBMBIDI
   nsContentUtils::RegisterPrefCallback("bidi.", PrefChangedCallback,
                                        this);
