diff -rwup mozilla-1.9.2.old/browser/components/preferences/advanced.xul mozilla-1.9.2/browser/components/preferences/advanced.xul
--- mozilla-1.9.2.old/browser/components/preferences/advanced.xul	2010-01-20 22:29:55.000000000 -0500
+++ mozilla-1.9.2/browser/components/preferences/advanced.xul	2010-03-26 13:27:05.000000000 -0400
@@ -120,6 +120,9 @@
       <preference id="security.disable_button.openDeviceManager"
                   name="security.disable_button.openDeviceManager"
                   type="bool"/>
+      <preference id="security.dnssec.dnssecBehavior"
+                  name="security.dnssec.dnssecBehavior"
+                  type="int"/>
     </preferences>
     
 #ifdef HAVE_SHELL_SERVICE
@@ -137,6 +140,7 @@
         <tab id="networkTab" label="&networkTab.label;" helpTopic="prefs-advanced-network"/>
         <tab id="updateTab" label="&updateTab.label;"  helpTopic="prefs-advanced-update"/>
         <tab id="encryptionTab" label="&encryptionTab.label;" helpTopic="prefs-advanced-encryption"/>
+        <tab id="dnssecTab" label="&dnssecTab.label;" helpTopic="prefs-advanced-dnssec"/>
       </tabs>
 
       <tabpanels flex="1">
@@ -397,6 +401,20 @@
           </groupbox>
         </tabpanel>
 
+        <tabpanel orient="vertical">
+          <groupbox>
+            <caption label="&dnsSecurity.label;"/>
+            <description>&dnssec.description;</description>
+           <radiogroup id="networkDnssecBehaviour"
+                       preference="security.dnssec.dnssecBehavior">
+             <radio value="1" label="&neverDNSSecure.label;"
+                    accesskey="&neverDNSSecure.accesskey;"/>
+             <radio value="0" label="&possibleDNSSecure.label;"
+                    accesskey="&possibleDNSSecure.accesskey;"/>
+           </radiogroup>
+          </groupbox>
+        </tabpanel>
+
       </tabpanels>
     </tabbox>
   </prefpane>
diff -rwup mozilla-1.9.2.old/browser/locales/en-US/chrome/browser/preferences/advanced.dtd mozilla-1.9.2/browser/locales/en-US/chrome/browser/preferences/advanced.dtd
--- mozilla-1.9.2.old/browser/locales/en-US/chrome/browser/preferences/advanced.dtd	2010-01-20 22:29:55.000000000 -0500
+++ mozilla-1.9.2/browser/locales/en-US/chrome/browser/preferences/advanced.dtd	2010-03-26 13:27:07.000000000 -0400
@@ -99,3 +99,13 @@
 <!ENTITY verify2.accesskey               "V">
 <!ENTITY viewSecurityDevices.label       "Security Devices">
 <!ENTITY viewSecurityDevices.accesskey   "y">
+
+<!ENTITY dnssecTab.label                "DNS Security">
+<!ENTITY dnssec.description             "DNS Security Policy is currently set at the system level (e.g. /etc/dnsval.conf) and this option specifies if the configured DNS Security policy should be enforced.">
+<!ENTITY dnsSecurity.label              "DNS Security Policy">
+<!ENTITY neverDNSSecure.label           "Ignore DNS security">
+<!ENTITY neverDNSSecure.accesskey       "N">
+<!ENTITY possibleDNSSecure.label        "Enforce DNS Security">
+<!ENTITY possibleDNSSecure.accesskey    "E">
+<!ENTITY allDNSSecure.label             "Require all DNS lookups to be secure">
+<!ENTITY allDNSSecure.accesskey         "A">
diff -rwup mozilla-1.9.2.old/browser/locales/en-US/chrome/overrides/appstrings.properties mozilla-1.9.2/browser/locales/en-US/chrome/overrides/appstrings.properties
--- mozilla-1.9.2.old/browser/locales/en-US/chrome/overrides/appstrings.properties	2010-01-20 22:29:55.000000000 -0500
+++ mozilla-1.9.2/browser/locales/en-US/chrome/overrides/appstrings.properties	2010-03-26 13:27:07.000000000 -0400
@@ -37,6 +37,8 @@
 malformedURI=The URL is not valid and cannot be loaded.
 fileNotFound=Firefox can't find the file at %S.
 dnsNotFound=Firefox can't find the server at %S.
+dnsDNSSECValidationError=The address (URL) failed its DNSSEC security check validation.  Please contact a system administrator for help.
+dnsDNSSECDNEError=The browser verified that there is no server for the provided address.
 protocolNotFound=Firefox doesn't know how to open this address, because the protocol (%S) isn't associated with any program.
 connectionFailure=Firefox can't establish a connection to the server at %S.
 netInterrupt=The connection to %S was interrupted while the page was loading.
diff -rwup mozilla-1.9.2.old/browser/locales/en-US/chrome/overrides/netError.dtd mozilla-1.9.2/browser/locales/en-US/chrome/overrides/netError.dtd
--- mozilla-1.9.2.old/browser/locales/en-US/chrome/overrides/netError.dtd	2010-01-20 22:29:55.000000000 -0500
+++ mozilla-1.9.2/browser/locales/en-US/chrome/overrides/netError.dtd	2010-03-26 13:27:07.000000000 -0400
@@ -25,6 +25,12 @@
 </ul>
 ">
 
+<!ENTITY dnsDNSSECValidationError.title "DNSSEC Validation Error">
+<!ENTITY dnsDNSSECValidationError.longDesc "The address (URL) failed its DNSSEC security check validation.  Please contact a system administrator for help.">
+
+<!ENTITY dnsDNSSECDNEError.title "Server Does Not Exist">
+<!ENTITY dnsDNSSECDNEError.longDesc "<p>The browser verified that there is no server for the provided address.</p><ul><li>Did you make a mistake when typing the domain? (e.g. <q><strong>ww</strong>.mozilla.org</q> instead of <q><strong>www</strong>.mozilla.org</q>)</li><li>This domain address does not exist.  Are you sure it is supposed to?  Maybe it used to but it no longer does?</li></ul>">
+
 <!ENTITY fileNotFound.title "File not found">
 <!ENTITY fileNotFound.longDesc "
 <ul>
diff -rwup mozilla-1.9.2.old/docshell/resources/content/netError.xhtml mozilla-1.9.2/docshell/resources/content/netError.xhtml
--- mozilla-1.9.2.old/docshell/resources/content/netError.xhtml	2010-01-20 22:30:06.000000000 -0500
+++ mozilla-1.9.2/docshell/resources/content/netError.xhtml	2010-03-26 13:27:36.000000000 -0400
@@ -294,6 +294,8 @@
     <div id="errorContainer">
       <div id="errorTitlesContainer">
         <h1 id="et_generic">&generic.title;</h1>
+        <h1 id="et_dnsDNSSECValidationError">&dnsDNSSECValidationError.title;</h1>
+        <h1 id="et_dnsDNSSECDNEError">&dnsDNSSECDNEError.title;</h1>
         <h1 id="et_dnsNotFound">&dnsNotFound.title;</h1>
         <h1 id="et_fileNotFound">&fileNotFound.title;</h1>
         <h1 id="et_malformedURI">&malformedURI.title;</h1>
@@ -317,6 +319,8 @@
       <div id="errorDescriptionsContainer">
         <div id="ed_generic">&generic.longDesc;</div>
         <div id="ed_dnsNotFound">&dnsNotFound.longDesc;</div>
+        <div id="ed_dnsDNSSECValidationError">&dnsDNSSECValidationError.longDesc;</div>
+        <div id="ed_dnsDNSSECDNEError">&dnsDNSSECDNEError.longDesc;</div>
         <div id="ed_fileNotFound">&fileNotFound.longDesc;</div>
         <div id="ed_malformedURI">&malformedURI.longDesc;</div>
         <div id="ed_protocolNotFound">&protocolNotFound.longDesc;</div>
diff -rwup mozilla-1.9.2.old/dom/locales/en-US/chrome/appstrings.properties mozilla-1.9.2/dom/locales/en-US/chrome/appstrings.properties
--- mozilla-1.9.2.old/dom/locales/en-US/chrome/appstrings.properties	2010-01-20 22:30:06.000000000 -0500
+++ mozilla-1.9.2/dom/locales/en-US/chrome/appstrings.properties	2010-03-26 13:27:39.000000000 -0400
@@ -37,6 +37,8 @@
 malformedURI=The URL is not valid and cannot be loaded.
 fileNotFound=The file %S cannot be found. Please check the location and try again.
 dnsNotFound=%S could not be found. Please check the name and try again.
+dnsDNSSECValidationError=%S failed its DNSSEC security check validation. Please contact a security or system administrator for help.
+dnsDNSSECDNEError=%S does not exist within the DNS.  Please check the name and try again.
 protocolNotFound=%S is not a registered protocol.
 connectionFailure=The connection was refused when attempting to contact %S.
 netInterrupt=The connection to %S has terminated unexpectedly. Some data may have been transferred.
diff -rwup mozilla-1.9.2.old/dom/locales/en-US/chrome/netError.dtd mozilla-1.9.2/dom/locales/en-US/chrome/netError.dtd
--- mozilla-1.9.2.old/dom/locales/en-US/chrome/netError.dtd	2010-01-20 22:30:06.000000000 -0500
+++ mozilla-1.9.2/dom/locales/en-US/chrome/netError.dtd	2010-03-26 13:27:39.000000000 -0400
@@ -12,6 +12,12 @@
 <!ENTITY dnsNotFound.title "Address Not Found">
 <!ENTITY dnsNotFound.longDesc "<p>The browser could not find the host server for the provided address.</p><ul><li>Did you make a mistake when typing the domain? (e.g. <q><strong>ww</strong>.mozilla.org</q> instead of <q><strong>www</strong>.mozilla.org</q>)</li><li>Are you certain this domain address exists?  Its registration may have expired.</li><li>Are you unable to browse other sites?  Check your network connection and DNS server settings.</li><li>Is your computer or network protected by a firewall or proxy?  Incorrect settings can interfere with Web browsing.</li></ul>">
 
+<!ENTITY dnsDNSSECValidationError.title "DNSSEC Validation Error">
+<!ENTITY dnsDNSSECValidationError.longDesc "The address (URL) failed its DNSSEC security check validation.  Please contact a system administrator for help.">
+
+<!ENTITY dnsDNSSECDNEError.title "Server Does Not Exist">
+<!ENTITY dnsDNSSECDNEError.longDesc "<p>The browser verified that there is no server for the provided address.</p><ul><li>Did you make a mistake when typing the domain? (e.g. <q><strong>ww</strong>.mozilla.org</q> instead of <q><strong>www</strong>.mozilla.org</q>)</li><li>This domain address does not exist.  Are you sure it is supposed to?  Maybe it used to but it no longer does?</li></ul>">
+
 <!ENTITY fileNotFound.title "File Not Found">
 <!ENTITY fileNotFound.longDesc "<ul><li>Could the item have been renamed, removed, or relocated?</li><li>Is there a spelling, capitalization, or other typographical error in the address?</li><li>Do you have sufficient access permissions to the requested item?</li></ul>">
 
diff -rwup mozilla-1.9.2.old/layout/base/nsPresContext.cpp mozilla-1.9.2/layout/base/nsPresContext.cpp
--- mozilla-1.9.2.old/layout/base/nsPresContext.cpp	2010-01-20 22:30:30.000000000 -0500
+++ mozilla-1.9.2/layout/base/nsPresContext.cpp	2010-03-26 13:28:38.000000000 -0400
@@ -70,6 +70,7 @@
 #include "nsIWeakReferenceUtils.h"
 #include "nsCSSRendering.h"
 #include "prprf.h"
+#include "prnetdb.h"
 #include "nsContentPolicyUtils.h"
 #include "nsIDOMDocument.h"
 #include "nsAutoPtr.h"
@@ -294,6 +295,11 @@ nsPresContext::~nsPresContext()
   nsContentUtils::UnregisterPrefCallback("image.animation_mode",
                                          nsPresContext::PrefChangedCallback,
                                          this);
+#ifdef MOZ_DNSSEC
+  nsContentUtils::UnregisterPrefCallback("security.dnssec.dnssecBehavior",
+                                         nsPresContext::PrefChangedCallback,
+                                         this);
+#endif
 #ifdef IBMBIDI
   nsContentUtils::UnregisterPrefCallback("bidi.", PrefChangedCallback, this);
 
@@ -712,6 +718,10 @@ nsPresContext::GetUserPreferences()
   else // dynamic change to invalid value should act like it does initially
     mImageAnimationModePref = imgIContainer::kNormalAnimMode;
 
+#ifdef MOZ_DNSSEC
+  PR_set_dnssec_validate_policy(nsContentUtils::GetIntPref("security.dnssec.dnssecBehavior"));
+#endif /* MOZ_DNSSEC */
+
   PRUint32 bidiOptions = GetBidi();
 
   PRInt32 prefInt =
@@ -893,6 +903,11 @@ nsPresContext::Init(nsIDeviceContext* aD
   nsContentUtils::RegisterPrefCallback("image.animation_mode",
                                        nsPresContext::PrefChangedCallback,
                                        this);
+#ifdef MOZ_DNSSEC
+  nsContentUtils::RegisterPrefCallback("security.dnssec.dnssecBehavior",
+                                       nsPresContext::PrefChangedCallback,
+                                       this);
+#endif
 #ifdef IBMBIDI
   nsContentUtils::RegisterPrefCallback("bidi.", PrefChangedCallback,
                                        this);
