--- firefox.spec.orig	2009-12-06 22:08:57.000000000 -0500
+++ firefox.spec	2009-12-09 13:45:13.000000000 -0500
@@ -1,3 +1,5 @@
+%define dist .1.dnssec
+%define _default_patch_fuzz 2
 %define homepage http://start.fedoraproject.org/
 %define default_bookmarks_file %{_datadir}/bookmarks/default-bookmarks.html
 %define firefox_app_id \{ec8030f7-c20a-464f-9b0e-13a3a9e97384\}
@@ -5,7 +7,7 @@
 %define mozappdir            %{_libdir}/%{name}-%{internal_version}
 %define tarballdir mozilla-1.9.1
 
-%define xulrunner_version 1.9.1.5-1
+%define xulrunner_version 1.9.1.5-1.1.dnssec
 %define internal_version %{version}
 
 %define official_branding    1
@@ -44,6 +46,11 @@
 
 # Upstream patches
 
+# DNSSEC Patches
+Patch201:       dnssec-firefox.patch
+Patch203:       dnssec-both.patch
+Patch205:       dnssec-mozconfig.patch
+
 %if %{official_branding}
 # Required by Mozilla Corporation
 
@@ -100,6 +107,19 @@
 echo "ac_add_options --with-libxul-sdk=\
 `pkg-config --variable=sdkdir libxul`" >> .mozconfig
 
+###############################
+# begin dnssec related patches
+%patch201 -b.dnssec -p1
+%patch203 -b.dnssec -p1
+%patch205 -b.dnssec -p0
+
+# rebuild configure(s) due to dnssec patches
+/bin/rm -f ./configure
+/usr/bin/autoconf-2.13 
+# end dnssec related patches
+###############################
+
+
 #---------------------------------------------------------------------
 
 %build
@@ -107,7 +127,7 @@
 
 # Mozilla builds with -Wall with exception of a few warnings which show up
 # everywhere in the code; so, don't override that.
-MOZ_OPT_FLAGS=$(echo $RPM_OPT_FLAGS | %{__sed} -e 's/-Wall//')
+MOZ_OPT_FLAGS=$(echo $RPM_OPT_FLAGS | %{__sed} -e 's/-Wall//' -e 's/-O2/-O0/' -e 's/-fexceptions//')
 export CFLAGS=$MOZ_OPT_FLAGS
 export CXXFLAGS=$MOZ_OPT_FLAGS
 
@@ -330,6 +350,9 @@
 #---------------------------------------------------------------------
 
 %changelog
+* Thu Nov  25 2009 Robert.Story@cobham.com - 3.5.5-1.dnssec.1
+- DNSSEC patches
+
 * Thu Nov  5 2009 Jan Horak <jhorak@redhat.com> - 3.5.5-1
 - Update to 3.5.5
 
