--- xulrunner.spec.orig	2009-12-06 22:08:58.000000000 -0500
+++ xulrunner.spec	2009-12-07 14:36:10.000000000 -0500
@@ -1,5 +1,7 @@
+%define dist .1.dnssec
+%define _default_patch_fuzz 2
 # Minimal required versions
-%define nspr_version 4.8
+%define nspr_version 4.8.2-1.1.dnssec
 %define nss_version 3.12.3.99
 %define cairo_version 1.6.0
 %define freetype_version 2.1.9
@@ -37,6 +39,10 @@
 # Upstream patches
 Patch100:       mozilla-ps-pdf-simplify-operators.patch
 
+# DNSSEC Patches
+Patch201:       dnssec-firefox.patch
+Patch203:       dnssec-both.patch
+
 # ---------------------------------------------------
 
 BuildRoot:      %(mktemp -ud %{_tmppath}/%{name}-%{version}-%{release}-XXXXXX)
@@ -165,6 +171,17 @@
 %{__rm} -f .mozconfig
 %{__cp} %{SOURCE10} .mozconfig
 
+###############################
+# begin dnssec related patches
+%patch201 -p1 -b .dnssec
+%patch203 -p1 -b .dnssec
+
+# rebuild configure(s) due to dnssec patches
+/bin/rm -f ./configure
+/usr/bin/autoconf-2.13
+# end dnssec related patches
+###############################
+
 #---------------------------------------------------------------------
 
 %build
