--- lftp-4.3.8-redhat/configure.ac	2012-07-03 07:01:16.000000000 -0400
+++ lftp-4.3.8/configure.ac	2012-08-28 14:26:16.463152154 -0400
@@ -298,7 +298,7 @@
     if test "$ac_cv_header_validator_validator_h" != yes; then
         AC_MSG_ERROR(Can't find validator.h (from dnssec-tools))
     fi
-    AC_CHECK_LIB(ssl, SHA1_Init)
+    AC_SEARCH_LIBS([SHA1_Init], [crypto ssl])
     AC_CHECK_LIB(sres, query_send)
     if test "$ac_cv_lib_sres_query_send" != yes; then
         AC_MSG_ERROR(Can't find libsres (from dnssec-tools))
